compat/technic_networks.lua

```diff
@@ -0,0 +1,125 @@
+
+local vecadd = vector.add
+local poshash = minetest.hash_node_position
+local get_node = minetest.get_node
+local cables = technic and technic.cables
+local machines = technic and technic.machine_tiers
+
+-- Check for technic mod compatibility, compatible with technic plus
+if not technic or not technic.remove_network or not technic.networks or not cables or not machines then
+	minetest.log("warning", "[jumpdrive] Incompatible technic mod loaded")
+	jumpdrive.technic_network_compat = function() end
+	return
+end
+
+-- Function to move technic network to another position
+local network_node_arrays = {"PR_nodes","BA_nodes","RE_nodes"}
+local function move_network(net, delta)
+	for _,tblname in ipairs(network_node_arrays) do
+		local tbl = net[tblname]
+		for i=#tbl,1,-1 do
+			tbl[i] = vecadd(tbl[i], delta)
+		end
+	end
+	local new_net_id = poshash(vecadd(technic.network2pos(net.id), delta))
+	local new_all_nodes = {}
+	local all_nodes = net.all_nodes
+	for old_node_id, old_pos in pairs(all_nodes) do
+		local new_pos = vecadd(old_pos, delta)
+		local node_id = poshash(new_pos)
+		cables[old_node_id] = nil
+		cables[node_id] = new_net_id
+		new_all_nodes[node_id] = new_pos
+	end
+	net.all_nodes = new_all_nodes
+end
+
+local function vecadd2(a, b, c)
+	return {
+		x = a.x + b.x + c.x,
+		y = a.y + b.y + c.y,
+		z = a.z + b.z + c.z,
+	}
+end
+
+local function add_edge_cable_net(pos, size, delta, t)
+	-- Find connected positions outside edge of box from given position
+	-- NOTE: At source position first found would be enough but not at target
+	local x = pos.x == 0 and -1 or (pos.x == size.x and 1)
+	local y = pos.y == 0 and -1 or (pos.y == size.y and 1)
+	local z = pos.z == 0 and -1 or (pos.z == size.z and 1)
+	if x then
+		local v = vecadd2(pos, {x=x,y=0,z=0}, delta)
+		local net_id = cables[poshash(v)]
+		if net_id then t[net_id] = true end
+	end
+	if y then
+		local v = vecadd2(pos, {x=0,y=y,z=0}, delta)
+		local net_id = cables[poshash(v)]
+		if net_id then t[net_id] = true end
+	end
+	if z then
+		local v = vecadd2(pos, {x=0,y=0,z=z}, delta)
+		local net_id = cables[poshash(v)]
+		if net_id then t[net_id] = true end
+	end
+	return x, y, z
+end
+
+jumpdrive.technic_network_compat = function(source_pos1, source_pos2, target_pos1, delta_vector)
+
+	-- search results
+	local jump_networks = {} -- jumped fully contained networks
+	local edge_networks = {} -- networks crossing jumped area edge
+
+	-- search for networks in area
+	local size = vector.apply(vector.subtract(source_pos1, source_pos2),math.abs)
+	for x=0,size.x do
+		for y=0,size.y do
+			for z=0,size.z do
+				local local_pos = {x=x,y=y,z=z}
```

Comment:
Problem here is when large areas are jumped, this most probably is not even near most efficient method to check nodes in area.
Any suggestions for better way?

Reply:
Wuld it be better to find_nodes_in_area and include cable groups and machine group, if that is significantly faster way to get nodes then loop here could be a lot shorter especially for large areas.

Reply:
uhm, this does not sound really efficient, would that be another use-case in that https://github.com/mt-mods/jumpdrive/issues/47 could help with. It would need to be defined on every cable-definition (and possibly on the switching station):

```lua
minetest.register_node("my:node", {
 on_movenode = function(from_pos, to_pos)
 end
})
```

This way the nested for-loops could be replaced with some logic within the jumpdrive mod (one nested for-loop is better than 2 of them, right?).

Let me know if something like that would help and if the parameters are enough (from- and to-position) i think i have some time to implement this and rewrite the jump-compatibility functions (or even move it to the external mods)

Reply:
Yeah, thing is with R15 it is still well over 50% faster than rebuilding network. I think mostly because it is very rare to read any nodes from world (only around edges of box when there is also active network at location) but other than that it is pure math.

Reply:
> This way the nested for-loops could be replaced with some logic within the jumpdrive mod (one nested for-loop is better than 2 of them, right?).

Node move cannot be used alone, did you notice that this loop also checks taget and source around jumped area for possible network connections at target and cable cutting at source.
It could be used if there's information about location in jumped area so that connections across edge can be checked.

But it would work yes with few additions and possibly better than blindly looping through whole area.

Reply:
It would work with following:
* collection phase that could use on_movenode to save data for nodes that would be moved
* finalization phase that could be used to verify connections between collected nodes and outside area
   this would then decide if network can be moved or must be destroyed and rebuilt

Maybe there's other ways too but cant think of any, suggestions?

Networks that are fully contained can be moved without additional actions.
Unless there is networks connecting to them at target location outside of jumped box:
https://github.com/mt-mods/jumpdrive/blob/2087759fbf69b16d4e06a4a6b3f2938a4a2f5994/compat/technic_networks.lua#L51-L53

Reply:
This would provide a way to react on single nodes and to process/finalize the whole after a jump:

```lua
local jumped_cables = {}

minetest.register_node("technic:cable_xy", {
 on_movenode = function(from, to, additional_info)
  -- TODO: collect cable-positions here
  table.insert(jumped_cables, { from=from, to=to, edge=additional_info.edge })
 end
})

-- callback for jump finalization stage (for example savefile writes)
jumpdrive.register_after_jump(function(from_area, to_area)
  -- from_area = { pos1, pos2 }
  -- to_area = { pos1, pos2 }

  -- TODO: finalize (somehow)
  for _, entry in ipairs(jumped_cables) do
    -- magic
  end

  -- reset list for next jump
  jumped_cables = {}
end)
```

Reply:
so `additional_info.edge` would contain 1-3 edge directions / vector representing directions, that would work yes.
Most of logic from this PR could be moved to finalization and if `edge` provides vector for 90 degree directions relative to edge wall then that could also be skipped in node move implementation for networks.

I think that could be useful for some other things too that can depend on nearby stuff outside of source or target area.

Reply:
btw, tried with following. Removed nested loops and used 254 nodes instead of 29791 nodes.
**those nested loops were about 8 times faster**...
```lua
        local nodes = minetest.find_nodes_in_area(source_pos1, source_pos2, {
                "group:technic_machine",
                "group:technic_lv_cable",
                "group:technic_mv_cable",
                "group:technic_hv_cable",
        })

        -- search results
        local jump_networks = {} -- jumped fully contained networks
        local edge_networks = {} -- networks crossing jumped area edge

        -- search for networks in area
        local size = vector.apply(vector.subtract(source_pos1, source_pos2),math.abs)
        --for x=0,size.x do
                --for y=0,size.y do
                        --for z=0,size.z do
        for _, pos in ipairs(nodes) do
....
```

Reply:
> additional_info.edge would contain 1-3 edge directions / vector representing directions, that would work yes.

Hmm, it does not (yet) only a boolean value if it _is_ on the edge :yum:
But i guess returning a vector for the edge-position should be fairly easy, something along the lines of this:
```lua
-- only an edge to the lower/y direction
edge = { x=0, y=-1, z=0 }
-- on an upper corner
egde = { x=1, y=1, z=-1 }
```

Reply:
Could put up some simple test for this and if it still seems that nested loops using just pure math is better then maybe good option would be to implement actual network moving in technic mod instead and just call some `technic.move_network(source1, source2, target1)` from here.
But seems that needs  some testing, and possibility to register move callback for nodes will be useful no matter how network data moving will be actually implemented.